sycl/include/CL/sycl/INTEL/pipe_array.hpp

```diff
@@ -0,0 +1,57 @@
+//==--------------- pipe_array.hpp - SYCL pipe array --------*- C++ -*------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// ===--------------------------------------------------------------------=== //
+
+#pragma once
+
+#include <CL/sycl/INTEL/pipes.hpp>
+#include <utility>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace INTEL {
+
+namespace {
+template <size_t dim1, size_t... dims> struct VerifierDimLayer {
+  template <size_t idx1, size_t... idxs> struct VerifierIdxLayer {
+    static constexpr bool IsValid() {
+      return idx1 < dim1 &&
+             (VerifierDimLayer<dims...>::template VerifierIdxLayer<
+                 idxs...>::IsValid());
+    }
+  };
+};
+template <size_t dim> struct VerifierDimLayer<dim> {
+  template <size_t idx> struct VerifierIdxLayer {
+    static constexpr bool IsValid() { return idx < dim; }
+  };
+};
+} // namespace
+
+template <class Id, typename BaseTy, size_t depth, size_t... dims>
+struct PipeArray {
```

Comment:
This is not really an array... But I cannot find a better name.
`pipe_array` looks more similar to SYCL coding style, otherwise.